src/PowerShellEditorServices.Hosting/HostLogger.cs

```diff
@@ -0,0 +1,332 @@
+    /// there's no good reason to need objects rather than writing directly to the host.
+    /// </remarks>
+    internal class PSHostLogger : IObserver<(PsesLogLevel logLevel, string message)>
+    {
+        private readonly PSHostUserInterface _ui;
+
+        /// <summary>
+        /// Create a new PowerShell host logger.
+        /// </summary>
+        /// <param name="ui">The PowerShell host user interface object to log output to.</param>
+        public PSHostLogger(PSHostUserInterface ui)
+        {
+            _ui = ui;
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        public void OnError(Exception error)
+        {
+            OnNext((PsesLogLevel.Error, $"Error occurred while logging: {error}"));
+        }
+
+        public void OnNext((PsesLogLevel logLevel, string message) value)
+        {
+            switch (value.logLevel)
+            {
+                case PsesLogLevel.Diagnostic:
+                    _ui.WriteDebugLine(value.message);
+                    return;
+
+                case PsesLogLevel.Verbose:
+                    _ui.WriteVerboseLine(value.message);
+                    return;
+
+                case PsesLogLevel.Normal:
+                    _ui.WriteLine(value.message);
+                    return;
+
+                case PsesLogLevel.Warning:
+                    _ui.WriteWarningLine(value.message);
+                    return;
+
+                case PsesLogLevel.Error:
+                    _ui.WriteErrorLine(value.message);
+                    return;
+            }
+        }
+    }
+
+    internal class StreamLogger : IObserver<(PsesLogLevel logLevel, string message)>, IDisposable
+    {
+        public static StreamLogger CreateWithNewFile(string path)
+        {
+            var fileStream = new FileStream(
+                path,
+                FileMode.Create,
+                FileAccess.Write,
+                FileShare.Read,
+                bufferSize: 4096,
+                FileOptions.Asynchronous | FileOptions.SequentialScan);
+
+            return new StreamLogger(new StreamWriter(fileStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)));
+        }
+
+        private readonly StreamWriter _fileWriter;
+
+        // This cannot be a bool
+        // See https://stackoverflow.com/q/6164751
+        private int _hasCompleted;
+
+        private IDisposable _unsubscriber;
+
+        public StreamLogger(StreamWriter streamWriter)
+        {
+            _hasCompleted = 0;
+            streamWriter.AutoFlush = true;
+            _fileWriter = streamWriter;
+        }
+
+        public void OnCompleted()
+        {
+            // Ensure we only complete once
+            if (Interlocked.Exchange(ref _hasCompleted, 1) != 0)
+            {
+                return;
+            }
+
+            _unsubscriber.Dispose();
+            _fileWriter.Flush();
+            _fileWriter.Close();
+            _fileWriter.Dispose();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Move this 'Dispose' call into this class' own 'Dispose' method.](https://app.codacy.com/app/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=4579423)